src/Microsoft.JSInterop/JavaScriptRuntime/src/Microsoft.JSInterop.ts

```diff
@@ -0,0 +1,234 @@
+     * @returns A Function instance.
+     */
+    findJSFunction,
+
+    /**
+     * Invokes the specified synchronous JavaScript function.
+     *
+     * @param identifier Identifies the globally-reachable function to invoke.
+     * @param argsJson JSON representation of arguments to be passed to the function.
+     * @returns JSON representation of the invocation result.
+     */
+    invokeJSFromDotNet: (identifier: string, argsJson: string) => {
+      const result = findJSFunction(identifier).apply(null, parseJsonWithRevivers(argsJson));
+      return result === null || result === undefined
+        ? null
+        : JSON.stringify(result);
+    },
+
+    /**
+     * Invokes the specified synchronous or asynchronous JavaScript function.
+     *
+     * @param asyncHandle A value identifying the asynchronous operation. This value will be passed back in a later call to endInvokeJSFromDotNet.
+     * @param identifier Identifies the globally-reachable function to invoke.
+     * @param argsJson JSON representation of arguments to be passed to the function.
+     */
+    beginInvokeJSFromDotNet: (asyncHandle: number, identifier: string, argsJson: string): void => {
+      // Coerce synchronous functions into async ones, plus treat
+      // synchronous exceptions the same as async ones
+      const promise = new Promise<any>(resolve => {
+        const synchronousResultOrPromise = findJSFunction(identifier).apply(null, parseJsonWithRevivers(argsJson));
+        resolve(synchronousResultOrPromise);
+      });
+
+      // We only listen for a result if the caller wants to be notified about it
+      if (asyncHandle) {
+        // On completion, dispatch result back to .NET
+        // Not using "await" because it codegens a lot of boilerplate
+        promise.then(
+          result => getRequiredDispatcher().beginInvokeDotNetFromJS(0, 'Microsoft.JSInterop', 'DotNetDispatcher.EndInvoke', JSON.stringify([asyncHandle, true, result])),
+          error => getRequiredDispatcher().beginInvokeDotNetFromJS(0, 'Microsoft.JSInterop', 'DotNetDispatcher.EndInvoke', JSON.stringify([asyncHandle, false, formatError(error)]))
+        );
+      }
+    },
+
+    /**
+     * Receives notification that an async call from JS to .NET has completed.
+     * @param asyncCallId The identifier supplied in an earlier call to beginInvokeDotNetFromJS.
+     * @param success A flag to indicate whether the operation completed successfully.
+     * @param resultOrExceptionMessage Either the operation result or an error message.
+     */
+    endInvokeDotNetFromJS: (asyncCallId: string, success: boolean, resultOrExceptionMessage: any): void => {
+      const resultOrError = success ? resultOrExceptionMessage : new Error(resultOrExceptionMessage);
+      completePendingCall(parseInt(asyncCallId), success, resultOrError);
+    }
+  }
+
+  function parseJsonWithRevivers(json: string): any {
+    return json ? JSON.parse(json, (key, initialValue) => {
```

Comment:
Interesting... I wasn't aware this feature exists 👍

Reply:
Yeah, typescript language features are amazing... Each version it surprises me more! 👍

Reply:
Well I this case I think it's browser APIs - but yes, Typescript is amazing. 😻